### PR TITLE
Added `FieldDef<'Val, 'Res>` to support resolver changing middlewares

### PR DESCRIPTION
**Pull request by xperiandri**
In order to call `Define.Field<'Val>(....).WithAuthorizationPolicy("policy")` I need to know not only `'Res` which is an object on which `FieldDef<'Res>` is defined but also a `'Val` which is a field type.
However when I introduced `FieldDef<'Val, 'Res>` F# cannot determine the override between `Define.Object` that gets a list of field and a labmda that return a list of fields. So I had to rename `Define.Object` overrie to `Define.ObjectRec` and `Define.Interface` override to `Define.ObjectRec`.
I have not renamed `Define.InputObject` as it accepts a parameter of list of another interface

### CODE REVIEW

**Comment by GBirkel** [comment]
I'm afraid this one may be beyond my skill to review.  I'm not sure what the long-term consequences of changing `Object` to `ObjectRec` would be.

**Comment by xperiandri** [comment]
Anyway we do breaking changes in v2.0
So the main difference is that you need to call `Define.ObjectRec` to define recursive GraphQL type definitions.
Also there is a question if need to rename `Define.InputObject` override to `Define.InputObjectRec` to be consistent or leave it as is to be simple?

File: src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs [hunk 1]
```diff
@@ -665,7 +665,8 @@ module SchemaDefinitions =
               DeprecationReason = deprecationReason }
 
         /// <summary>
-        /// Creates GraphQL custom output object type. It can be used as a valid output but not an input object
+        /// Creates GraphQL custom output object type that has a field of type that refernces this object type recurively.
```
Comment (valbers) [inline src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs]:
```suggestion
        /// Creates a GraphQL custom output object type that has a field of a type that references this object type recursively.
```

File: src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs [hunk 2]
```diff
@@ -665,7 +665,8 @@ module SchemaDefinitions =
...
+        /// It can be used as a valid output but not an input object
```
Comment (valbers) [inline src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs]:
```suggestion
        /// It can be used as a valid output but not as an input object
```

**Review by valbers — CHANGES_REQUESTED** [review]
I get it.
In any case, I find it a good idea to generally avoid method overloads as F# doesn't support function overloads (although class members may have overloads, just like in our current situation -- which is causing the problem that you're trying to fix), and therefore not having overloads feels more like "the F# way", so I welcome this change already on this level, even if the offending overload were not causing any problems yet.

However, I'm not sure if I agree with the new name. `ObjectRec` is effectively not by definition recursive, but it _can_ be used for a recursive definition.
That, which `ObjectRec` actually is, is an `Object` which takes a function for defining the fields instead of a concrete list of fields.

So, following the nomenclature of the F# core lib, I'd prefer for it to be named something like the [`defaultWith`](https://fsharp.github.io/fsharp-core-docs/reference/fsharp-core-optionmodule.html#defaultWith) function of the `Option` module.

In that case, that could be `ObjectHavingFieldsWith` (instead of `ObjectRec`).
What do you think?

**Review by GBirkel — APPROVED** [review]
Have the changes requested by valbers been addressed?

**Comment by xperiandri** [comment]
@njlr give your opinion, please

**Comment by valbers** [comment]
@xperiandri could you please do less force-pushes with commit amendments? It makes it hard to follow what you actually changed with each commit amendment. I mean: once in a while it's ok, but I see this now very often.

File: src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs [hunk 3]
```diff
@@ -800,15 +801,14 @@ module SchemaDefinitions =
         /// <param name="description">Optional field description. Usefull for generating documentation.</param>
         /// <param name="args">Optional list of arguments used to parametrize field resolution.</param>
         /// <param name="deprecationReason">If set, marks current field as deprecated.</param>
-        static member AutoField(name : string, typedef : #OutputDef<'Res>, ?description: string, ?args: InputFieldDef list, ?deprecationReason: string) : FieldDef<'Val> =
+        static member AutoField(name : string, typedef : OutputDef<'Res>, ?description: string, ?args: InputFieldDef list, ?deprecationReason: string) : FieldDef<'Val, 'Res> =
```
Comment (Thorium) [inline src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs:804]: Was the # removed on purpose?

Comment (xperiandri) [reply src/FSharp.Data.GraphQL.Shared/SchemaDefinitions.fs:804]: I either got a warning or error because of this

**Review by valbers — APPROVED** [review]
I gave you my two cents with regard to the naming of `ObjectRec`, but it seems like you don't agree with it. I don't want to die on that hill, so I'm approving this PR.

**Comment by xperiandri** [comment]
I just want to check again if it still the issue.
And I cannot decide on naming 🙂

**Comment by xperiandri** [comment]
@valbers I implemented an alternative solution #492
